builds/jazz-build-module/vault-utility-module.groovy

```diff
@@ -0,0 +1,114 @@
+#!groovy?
+import groovy.json.JsonSlurperClassic
+import groovy.json.JsonOutput
+import groovy.transform.Field
+
+
+echo "Vault utility module loaded successfully"
+
+
+@Field def configLoader
+@Field def baseUrl
+@Field def serviceConfig
+@Field def authToken
+@Field def environmentLogicalId
+
+def initialize(config_loader,service_config, base_url, auth_token, env) {
+	configLoader = config_loader
+	serviceConfig = service_config
+	baseUrl = base_url
+	authToken = auth_token
+	environmentLogicalId = env
+}
+
+def updateCustomServicesSafeDetails(safeName, lambdaArns, credsId) {
+	for (arn in lambdaArns) {
+		updateSafeDetails(safeName, arn, credsId)
+	}
+}
+
+def updateSafeDetails(safeName, lambdaARN, credsId) {
+	def iamRoleArn	
+	def safeDetails = getSafeDetails(safeName)
+
+	if (safeDetails) {
+		iamRoleArn = getRoleDetails(lambdaARN, credsId)
+		if(safeDetails.data.roles && safeDetails.data.roles.length != 0) {
+			def isRoleArnExists = safeDetails.data.roles.find{it -> it.value.arn == iamRoleArn}
+			if(!isRoleArnExists) {
+				addRoleToSafe(iamRoleArn, safeName)
+			} else echo "Role already exists"			
+		} else {
+			addRoleToSafe(iamRoleArn, safeName)
+		}
+	} else {
+		echo "Safe not configured yet."
+	}
+
+	return iamRoleArn 
+}
+
+def addRoleToSafe(iamRoleArn, safeName) {
+	try {
+		def rolePayload = [
+			'arn': iamRoleArn,
+			'permission': 'read'
+		]
+
+		def payload = JsonOutput.toJson(rolePayload)
+
+		def vaultApi = "${baseUrl}/jazz/t-vault/safes/${safeName}/role"
+		def statusCode = sh(script: "curl -H \"Content-type: application/json\" \
+			-H \"Jazz-Service-ID: ${serviceConfig['service_id']}\" \
+			-H \"Authorization: $authToken \" -X POST \
+			--write-out '%{http_code}\n' --silent --output /dev/null \
+			-d \'${payload}\' \"${vaultApi}\" ", returnStdout: true).trim()
+
+		if(statusCode == '200') echo "Successfully added role ${iamRoleArn} to safe ${safeName}" 
+		else echo "Error in adding role ${iamRoleArn} to safe ${safeName}"
+	} catch (ex) {
+		echo "Error in getting safe details. ${ex}"
+	}
+}
+
+def getSafeDetails(safeName) {
+	def vaultApi = "${baseUrl}/jazz/t-vault/safes/${safeName}"
+	def safeResponce = sh(script: "curl -H \"Content-type: application/json\" \
+		-H \"Jazz-Service-ID: ${serviceConfig['service_id']}\" \
+		-H \"Authorization: $authToken \" \
+		-X GET \"${vaultApi}\" ", returnStdout: true).trim()
+
+	def safeDetails
+	 try {
+		safeDetails = parseJson(safeResponce)
+	} catch (ex) {
+		echo "Error in getting safe details. ${ex}"
+	}
+	return safeDetails
+}
+
+def getRoleDetails(lambdaARN, credsId) {
```

Comment:
It would be better if we just persist the role that we create for the environment in its metadata instead of making one more call to AWS lambda service.

Reply:
If user supplies custom role, we already have a handle of it during the deployment. We're fetching list of roles having access to the safe previously. Compare the user supplied role with roles attached to the safe and if we don't see it, just add it to the safe.

Reply:
I am getting the role details from the lambda to keep it consistent for api, function and sls-app. In the case of sls-app role arn is available only after the deployment. Also there is chance for different roles for each lambda in sls-app. Also in sls-app the role names are generated by the serverless if the user not specified.

Reply:
Since we can have multiple role arns per service,  should we persist them in catalog ? In order to track the jazz created roles we can use the asset table.